components/core/src/clp/ffi/ir_stream/utils.hpp

```diff
@@ -133,5 +216,86 @@ template <typename encoded_variable_t>
     }
     return succeeded;
 }
+
+template <IntegerType T>
+auto get_ones_complement(T int_val) -> T {
+    return static_cast<T>(~int_val);
+}
+
+template <bool is_auto_generated, int8_t length_indicator_tag, SignedIntegerType encoded_node_id_t>
+auto size_dependent_encode_and_serialize_schema_tree_node_id(
+        SchemaTree::Node::id_t node_id,
+        std::vector<int8_t>& output_buf
+) -> void {
+    output_buf.push_back(length_indicator_tag);
+    if constexpr (is_auto_generated) {
+        serialize_int(get_ones_complement(static_cast<encoded_node_id_t>(node_id)), output_buf);
+    } else {
+        serialize_int(static_cast<encoded_node_id_t>(node_id), output_buf);
+    }
```

Comment:
I think it's possible for the caller to pass the wrong type for the corresponding tag byte, right? I think we could resolve this by making this method take a list of valid tag bytes as well as the correct tag byte, then we compare and choose the corresponding cast.

Similar comment for the size-dependent deserialize function.

Reply:
discussed offline and we decide to use local templated lambda.